### PR TITLE
Temp Fix: Avoid Prior Jumps

### PR DESCRIPTION
**Pull request by Scienfitz**
This is supposed to be a mild fix to the fact that our current default prior logic (contained in the default kernel factory) exhibits strong jumps, depending in which regime one ends up in.

The updated logic 
- took the low and high dimensional regimes from the EDBO presets
- linearly interpolates the prior moments between these regimes
- scrapped the other two regimes, ie has no dependency on parameter types anymore, but only overall dimensionality
- accounts for reduction of the effective dimensions by subtracting the amount of task parameters
- the previous point was also introduced into the `EDBOKernelFactory`, to which the old default was renamed. Additionally, all previous changes to the ebdo logic were reverted to reflect the original

This hopefully achieves a `mild` and sort of `safe` change recovering the existing results, but avoiding the aforementioned jumps (that were indeed observed in some stakeholder projects).

Eventually, this is to be replaced by a potentially entirely new prior logic based on the recently reported sqrt scaling.

Here are two results I checked against heir published state on the docs (old on top, new below):
image
image

image
image

### CODE REVIEW

File: baybe/surrogates/gaussian_process/presets/default.py
Comment (AVHopp) [inline baybe/surrogates/gaussian_process/presets/default.py:83]: This is to similar with the docstring of the other factory, since the only word that is different in the headline is "default"

File: baybe/surrogates/gaussian_process/presets/default.py
Comment (AVHopp) [inline baybe/surrogates/gaussian_process/presets/default.py:196]: This code is contained 3 times in this file. Small utility?

**Review by AVHopp — APPROVED** [review]
Just some minor comments, mostly happy.

File: baybe/surrogates/gaussian_process/presets/default.py
Comment (Scienfitz) [reply baybe/surrogates/gaussian_process/presets/default.py:88]: hmmmm seems it doesnt recognize that page is correct, follow the link its fine

File: CHANGELOG.md
Comment (Scienfitz) [reply CHANGELOG.md:52]: it was more than just a renaming since in previous attempts fix fix/shift the problem we shifted one of the threshold, that has been reverted and now is fully original logic (hence the phrasing)

File: baybe/surrogates/gaussian_process/presets/default.py
Comment (Scienfitz) [reply baybe/surrogates/gaussian_process/presets/default.py:83]: what is your suggestion? Do we have any design guideline regarding `too similar`?

File: baybe/surrogates/gaussian_process/presets/default.py
Comment (Scienfitz) [reply baybe/surrogates/gaussian_process/presets/default.py:196]: open to that, prob best part of the searchspace?

File: baybe/surrogates/gaussian_process/presets/default.py
Comment (AVHopp) [reply baybe/surrogates/gaussian_process/presets/default.py:88]: Huh, weird. We could add this link as one that is ignored by `linkcheck` or not have the full link, which do you prefer?

File: baybe/surrogates/gaussian_process/presets/default.py
Comment (AVHopp) [reply baybe/surrogates/gaussian_process/presets/default.py:83]: `A factory providing kernels using a logic adapted from EDBO` or something like that?

File: baybe/surrogates/gaussian_process/presets/default.py
Comment (AVHopp) [reply baybe/surrogates/gaussian_process/presets/default.py:196]: Having a small `@property` for `effective_dims` feels like something that might be nice from both a user view (--> how much parameters do I have effectively?) and obviously also from a coding perspective.

File: baybe/surrogates/gaussian_process/presets/default.py
Comment (Scienfitz) [reply baybe/surrogates/gaussian_process/presets/default.py:82]: not tested

File: baybe/surrogates/gaussian_process/presets/default.py
Comment (Scienfitz) [reply baybe/surrogates/gaussian_process/presets/default.py:196]: this is kind of what it says directly in the line below so let me know if I should delete that because having it in both places feels weird

File: baybe/surrogates/gaussian_process/presets/default.py
Comment (Scienfitz) [reply baybe/surrogates/gaussian_process/presets/default.py:83]: this is kind of what it says directly in the line below so let me know if I should delete that because having it in both places feels weird

File: baybe/surrogates/gaussian_process/presets/default.py
Comment (Scienfitz) [reply baybe/surrogates/gaussian_process/presets/default.py:88]: should be ingored, let me know how to do that, I assume on conf.py somewhere? do we have a precedent?

File: baybe/surrogates/gaussian_process/presets/default.py
Comment (Scienfitz) [reply baybe/surrogates/gaussian_process/presets/default.py]: added

File: baybe/surrogates/gaussian_process/presets/default.py
Comment (Scienfitz) [reply baybe/surrogates/gaussian_process/presets/default.py:76]: added now, please check
kernel factory was not tested preivously afaik because the default kernel fixture is set to a plain matern kernel not factory

**Review by AdrianSosic — CHANGES_REQUESTED** [review]
Hi @Scienfitz, before I go into the review, can you do me a little favor and please move the EDBO stuff to a new preset file `edbo.py`? The whole point of the preset folder was to manage several completely isolated settings, i.e. in the ideal case the files should never be touched again once correctly implemented. Plus, it will make the review process / diffs easier 🙃

**Comment by Scienfitz** [comment]
@AdrianSosic split the files as requested

File: baybe/surrogates/gaussian_process/presets/default.py
Comment (AVHopp) [reply baybe/surrogates/gaussian_process/presets/default.py]: You should be able to achieve this by adding a `linkcheck_ignore` in the `conf.py`, see https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-linkcheck_ignore. We previously did that when certain parts of the documentation were not yet built but we knew that the run would build them (e.g. in the very beginning) and that worked as expected.

File: tests/test_iterations.py
Comment (AVHopp) [inline tests/test_iterations.py:243]: Just out of curiosity, why is the recommender being printed here?

File: baybe/surrogates/gaussian_process/presets/default.py
Comment (AVHopp) [reply baybe/surrogates/gaussian_process/presets/default.py]: I would prefer to have these two lines merged. But if you feel that this is unnecessary, just resolve here since this is really something minor.

File: tests/test_iterations.py
Comment (Scienfitz) [reply tests/test_iterations.py]: leftover, removed

File: baybe/surrogates/gaussian_process/presets/default.py
Comment (Scienfitz) [reply baybe/surrogates/gaussian_process/presets/default.py]: ok I removed the second line and put `EDBO` in the first line, this maxes our characters and no other content can go into that line

File: baybe/surrogates/gaussian_process/presets/default.py
Comment (Scienfitz) [reply baybe/surrogates/gaussian_process/presets/default.py]: fixed

File: baybe/surrogates/gaussian_process/presets/default.py
Comment (AdrianSosic) [inline baybe/surrogates/gaussian_process/presets/default.py:56]: I think you are reinventing the wheel here. Why not simply take `numpy.interp`? If you further pull out the prior float numbers first and interpolate them before creation the `GammaPrior` objects in our if-else blocks, you can even reuse a single `GammaPrior` call for each component, since `numpy.interp` by default keeps the values constant outside the specified boundaries. I think the final code will then have like 30% of its current length + no need to define custom interpolators.

File: baybe/searchspace/core.py
Comment (AdrianSosic) [inline baybe/searchspace/core.py:244]: TBH, I not convinced of these properties, at least not of their current form. The two problems I see here:
* We artificially couple our search space to concepts that lie outside its responsibilities, i.e. kernels and definitions like "default kernel" (<-- what is that even supposed to mean???), even if that's done only verbally.
* You are effectively moving the decision how to treat kernel dimensions (which should naturally belong to the surrogate) to the searchspace, which should not have to bother about these things at all. In particular, who says that all surrogates have to treat the task dimension separately? Currently, this is a very special setting ONLY for the GP process – other surrogate might do it differently, and even future versions of the GP (e.g. when introducing multi-fidelity etc.) could deviate.

So if you really think it's worth having these properties, which I don't reject in general, please let's rewrite them such that the terminology and the implied decisions remain completely neutral and decoupled from other classes. For example, instead of talking about kernel and stuff, on can simply stick to terminology that is known to the class (or classes below), such as task parameters. (perhaps: `comp_rep_n_non_task_dimensions` ?)

File: baybe/searchspace/continuous.py
Comment (AdrianSosic) [inline baybe/searchspace/continuous.py:489]: See my longer comment about these properties below, but for this one in particular I'm not sure if it's really needed, since I can't imagine someone specifically asking a `ContinuousSubspace` about infos on its `TaskParameters`, a concept which doesn't exist in conti spaces. Since everyone would access this info via `SearchSpace`, I feel that this property is unnecessary?

File: baybe/searchspace/discrete.py
Comment (AdrianSosic) [inline baybe/searchspace/discrete.py:243]: properties are much further down in the file

**Review by AdrianSosic — CHANGES_REQUESTED** [review]
Hi @Scienfitz, thx for taking care of it. Core changes are conceptually fine, just a few comments on how things are implemented 👍🏼

File: baybe/searchspace/core.py
Comment (Scienfitz) [reply baybe/searchspace/core.py:244]: the idea is based on the fact that we currently have one overreaching kernel type for all comp parameter columns, with the exception of the task parameter. I simply call this the default kernel and this property is supposed to compute how many dimensions would be affected (= all dims - task dims) - that this could not be the case in current code is news to me. 

If the future eg introduces parameter-specific kernels they could simply be subtracted here too (your name suggestion then wouldnt make sense tho). In this package, any talk about kernel implies GP, if a surrogate doesnt use the GP kernels, it wouldnt use this property

Its related to GP kernels so not ahving that int he name imo doesnt make sense. If your arg is about coupling then I'd rather remove this property, its not very needed. Let me know

File: baybe/searchspace/continuous.py
Comment (Scienfitz) [reply baybe/searchspace/continuous.py:489]: agree that taskparameter is not relevant here

But this was already put here with regards to consistency / the future: there could be various other mechanism that reduce the default kernel dimensions, e.g. if I specify that for 1 our of 4 conti parameters I want a special (parameter-specific) kernel, the effective default kernel dim would be 3 not 4, that would be reflected here.

Comment (AdrianSosic) [reply baybe/searchspace/continuous.py:489]: Yes, but I think that only confirms my argument: I attribute this decision to the surrogate, *not* to the searchspace. The searchspace is not aware how any of its components are to be handled and should thus not decide it. Different surrogates can apply arbitrarily different rules here. So in essence I suggest: if you want to extract the code into a functionality, it should belong the class making use of it / defining these rules. And that wouldn't even be the general surrogate class but only the *specific* surrogate classes that implement this particular split. Do you agree with this argument?

Thus, in your case, I could imagine it as a utility defined somewhere close to our current GP model, i.e. a simple function that takes the searchspace as input. But I'll leave it to you do decide if this improves the code or not.

File: baybe/searchspace/core.py
Comment (AdrianSosic) [reply baybe/searchspace/core.py:244]: Yes, it's about decoupling and perhaps even more about which object is responsible for making the decision how the dimensions are to be divided (see answer in other thread)

File: baybe/searchspace/discrete.py
Comment (Scienfitz) [reply baybe/searchspace/discrete.py]: property removed

File: baybe/searchspace/continuous.py
Comment (Scienfitz) [reply baybe/searchspace/continuous.py]: property removed

File: baybe/searchspace/core.py
Comment (Scienfitz) [reply baybe/searchspace/core.py]: property removed

File: baybe/surrogates/gaussian_process/presets/default.py
Comment (Scienfitz) [reply baybe/surrogates/gaussian_process/presets/default.py]: forgot this exists, added it

File: baybe/surrogates/gaussian_process/presets/default.py
Comment (AdrianSosic) [inline baybe/surrogates/gaussian_process/presets/default.py:22]: In the remaining code base, we put these as docstrings below the constants

File: baybe/surrogates/gaussian_process/presets/default.py
Comment (AdrianSosic) [inline baybe/surrogates/gaussian_process/presets/default.py:45]: I'm not sure from where exactly you took the upper values. Seem like average of mordred and OHE, is that correct? If so, could you add corresponding comments here and to the noise factory?

**Review by AdrianSosic — CHANGES_REQUESTED** [review]
Thx for the changes! Here two remaining things before merge

File: baybe/surrogates/gaussian_process/presets/default.py
Comment (Scienfitz) [reply baybe/surrogates/gaussian_process/presets/default.py]: ok, changed
why do we do that again?

File: baybe/surrogates/gaussian_process/presets/default.py
Comment (Scienfitz) [reply baybe/surrogates/gaussian_process/presets/default.py:46]: exactly right, added

File: baybe/surrogates/gaussian_process/presets/default.py
Comment (AdrianSosic) [reply baybe/surrogates/gaussian_process/presets/default.py]: I think
* For consistency (in our classes, we also put the docstring below attributes)
* When they are public, they will actually appear in the doc, I think

Comment (AdrianSosic) [reply baybe/surrogates/gaussian_process/presets/default.py]: Just noticed:
* The "comment" below the variable should really be a docstring, i.e. using triplet quotes
* Please add the note about the average also to the noise factory, because its an isolated piece of code

**Review by AdrianSosic — APPROVED** [review]
Reopen that one issue that still needs to be addressed, but other than that, ready for merge, so already accepting

File: baybe/surrogates/gaussian_process/presets/default.py
Comment (Scienfitz) [reply baybe/surrogates/gaussian_process/presets/default.py:22]: I'm not sure you're right about the constants, where do we actually do this? This case here is `private` and `not a class attribute`, so why would anything you said apply? In fact looking for isntance here 
image
 I dont see the comment i) below ii) in triple quotes. Why did you suggest this? I think I should actually reverse it and have the comment above for consistency

Added the comment to the noise factory

Comment (AdrianSosic) [reply baybe/surrogates/gaussian_process/presets/default.py:22]: The case you showed is an exception since not user-facing and indeed part of a private package `_optional` that does not appear in the docs. And the thing put as comment is not describing the individual meaning of these constants but rather takes the role as a heading for a code block.

We do in fact have a few rare examples, like `Smiles`:
https://emdgroup.github.io/baybe/_autosummary/baybe.parameters.substance.html#module-baybe.parameters.substance

As you can see, the point is that things will get rendered in the doc. So even if the variable in your case is private, it makes sense to use a docstring:
* for consistency
* because it's 100% guaranteed that we would forget to turn the comment into a docstring in cases where we turn a private constant in a public one during refactoring

Comment (AdrianSosic) [reply baybe/surrogates/gaussian_process/presets/default.py:22]: I see that you've simply merged without actually having us resolved the issue 😞 ...

Agree that this erroneously public and I'm not sure if there are other actual examples in the current doc, but that is not the point of my argument! The rationale is simply: Everything public needs a docstring. Why use different rules for private objects? Makes it more complicated + they can become public in the future! Also, according to your logic, we should also put no docstrings under our private attributes!

### Your logic (not consistent, requires changes during refactoring)
```python
PUBLIC_CONSTANT = 1
"""Here you agree with docstring."""

# Following your logic, i.e. since not rendered, we'd need to put a comment here. 
# As soon as we make this value public, though, we need to turn the comment into a docstring,
# with additional effort and high chance of forgetting
_PRIVATE_CONSTANT = 1

@define
class A:
    public_attr = 0
    """Here you agree with docstring."""

    # Following your logic, i.e. since not rendered, we'd need to put a comment here. 
    # As soon as we make this value public, though, we need to turn the comment into a docstring,
    # with additional effort and high chance of forgetting
    _private_attr = 0
```
 
### My logic (everything consistent, no need for any changes)
```python
PUBLIC_CONSTANT = 1
"""Docstring."""

_PRIVATE_CONSTANT = 1
"""Docstring."""

@define
class A:
    public_attr = 0
    """Docstring."""

    _private_attr = 0
    """Docstring."""
```

**Pull request merged**
